### ISSUE TITLE
`11b_missing_observations_latent_path_mcmc.ipynb` runs very slowly on an Apple M4 Pro laptop

### ISSUE THREAD

**Issue by juanitorduz** [BasisResearch/dynestyx#353]
Follow-up to https://github.com/BasisResearch/dynestyx/pull/346#issuecomment-5465219164, where @DanWaxman asked for machine/environment details after I reported the `11b` notebook taking more than 2 hours while it takes ~1 minute end-to-end on his laptop.

## Summary

Running `docs/tutorials/gentle_intro/11b_missing_observations_latent_path_mcmc.ipynb` (branch `feature/arviz-1-migration`, commit `af0e560`, which already includes Dan's rerun from juanitorduz/dynestyx#3) on my machine is far slower than the ~1 minute reported. Per-cell timings below.

## Per-cell timings (2026-09-02)

I executed the notebook cell by cell with a 300 s cap per cell:

| Cell | Content | Wall time |
|---|---|---|
| imports | `import arviz as az ...` | 2.7 s |
| constants / helpers | `obs_times = jnp.arange(...)` | 0.1 s |
| `ar1_model` + synthetic data + plots | `def ar1_model(...)` | 1.4 s |
| **first MCMC cell** (`NUTS(conditioned_latent_path)`, 150 warmup + 150 samples, then `NUTS(conditioned_smoother)`) | `def conditioned_latent_path(...)` | **> 300 s, aborted** |

So the whole notebook up to the first `mcmc_latent_path.run(...)` takes ~4 s, and the very first latent-path NUTS run alone already exceeds the ~1 minute reported for the full notebook. The rest of the notebook (the `asymmetric_laplace` cell uses 10x more warmup/samples) was not reached. This is consistent with my earlier full run, which I killed after more than 2 hours. An earlier `nbconvert --execute` attempt today was also still inside this same cell after 4 minutes.

The notebook itself is unchanged in the MCMC settings vs `main` (`n_mcmc_warmup = n_mcmc_samples = 150`); only the ArviZ plotting cells differ.

## Machine

| | |
|---|---|
| Model | MacBook Pro (Mac16,8), Apple M4 Pro |
| Cores | 14 (10 performance + 4 efficiency) |
| Memory | 24 GB |
| OS | macOS 26.6.2 (build 25G83), arm64, no Rosetta (`sysctl.proc_translated = 0`) |
| Power | on battery during the run, Low Power Mode off, no thermal/performance warnings recorded (`pmset -g therm`) |

## Python environment

Created with `uv sync` from the repo `uv.lock` (`uv pip check`: all 218 packages compatible; installed versions match the lockfile).

| Package | Version |
|---|---|
| Python | 3.12.12 (uv-managed CPython, `cpython-3.12.12-macos-aarch64-none`) |
| uv | 0.12.9 |
| jax / jaxlib | 0.11.1 / 0.11.1 (wheel tag `cp312-cp312-macosx_11_0_arm64`) |
| numpyro | 0.21.0 |
| numpy | 2.5.2 (BLAS: Apple Accelerate) |
| scipy | 1.18.1 |
| arviz / arviz-base / arviz-stats / arviz-plots | 1.3.0 / 1.3.0 / 1.3.1 / 1.3.1 |
| blackjax | 1.6.2 |
| cuthbert / cuthbertlib | 0.0.14 / 0.0.14 |
| diffrax | 0.7.2 |
| equinox | 0.13.8 |
| tfp-nightly | 0.26.0.dev20260828 |
| ml-dtypes | 0.6.0 |
| dynestyx | 0.4.0 (editable install of the branch) |

JAX runtime: `jax.devices() == [CpuDevice(id=0)]`, `default_backend() == "cpu"`, `jax_enable_x64 = False`. No `XLA_FLAGS`, `JAX_*`, `OMP_NUM_THREADS`, or BLAS env vars are set in the shell or the notebook.

## How I ran it

Headless, with the project venv kernel, from the notebook's directory, on an otherwise idle machine (only browser + IDE open):

```bash
uv sync
uv run jupyter nbconvert --to notebook --execute \
  --ExecutePreprocessor.timeout=-1 \
  docs/tutorials/gentle_intro/11b_missing_observations_latent_path_mcmc.ipynb
```

For the per-cell numbers above I drove the same notebook with `nbclient` cell by cell and capped each cell at 300 s so I would not wait hours again. The kernel process sits at ~100% CPU (a single core) for the whole run.

## Questions / things that might help narrow this down

- Could you share `jax`, `jaxlib`, `numpyro`, and `cuthbert` versions plus Python version from the environment where it took ~1 min? If the lockfile on `main` resolves to different versions on your side, that would be the first thing I'd compare.
- Happy to run any diagnostic script you suggest (e.g. timing a single `NUTS` step or the compile of `conditioned_latent_path`) and report back.

**Comment by DanWaxman** [comment]
Thanks for the report @juanitorduz! I did some investigation with Claude on this, and it seems `jax>=0.11.0` has a regression in `jnp.stack` (or, rather, `jnp.concatenate`) and it might've been our postprocessing in that notebook, rather than MCMC. Does #354 fix things on your end? (If not, could you try using [my `uv.lock`](https://gist.github.com/DanWaxman/7f182d1380ef779a433049e31ff2900d)?)

**Comment by DanWaxman** [comment]
#354 landed, and at least fixes this problem to the extent that it is reproducible on my end -- @juanitorduz I'll close this for now, if it doesn't actually fix things on your side let me know and I'll re-open :) thanks again for the detailed bug report!!
